### PR TITLE
socket: unblock pending Read(), if any.

### LINKED ISSUE #17: Is it possible to "free" the physical Bluetooth interface so that it can be used by external commands?
I would like to be able to use the physical Bluetooth interface without stopping the go application so I need some way to free up the device once it has been initialised. Is this possible?

While the code is running I get errors as expected:
```
package main

import (
	"flag"
	"fmt"
	"log"
	"time"

	"github.com/currantlabs/ble/examples/lib/dev"
)

var (
	device = flag.String("device", "default", "implementation of ble")
)

func main() {
	_, err := dev.NewDevice(*device)
	if err != nil {
		log.Fatalf("can't new device : %s", err)
	}
	fmt.Println("opened")

	time.Sleep(100 * time.Second)
}
```
```
 ~  gatttool -I -t random -b E0:1A:35:05:47:EB
[E0:1A:35:05:47:EB][LE]> connect
Attempting to connect to E0:1A:35:05:47:EB
Error: connect: No route to host (113)
[E0:1A:35:05:47:EB][LE]> 
[E0:1A:35:05:47:EB][LE]> 
 ~  sudo hciconfig hci0 reset
Can't down device hci0: Device or resource busy (16)
 ✘  ~  sudo hciconfig hci0 down 
Can't down device hci0: Device or resource busy (16)
 ✘  ~  hciconfig dev
hci0:    Type: Primary  Bus: USB
    BD Address: E4:B3:18:4D:40:8F  ACL MTU: 1021:4  SCO MTU: 96:6
    UP RUNNING 
    RX bytes:49338 acl:2 sco:0 events:3888 errors:0
    TX bytes:625918 acl:181 sco:0 commands:4087 errors:0

 ~  gatttool -I -t random -b E0:1A:35:05:47:EB
[E0:1A:35:05:47:EB][LE]> connect
Attempting to connect to E0:1A:35:05:47:EB
Error: connect: No route to host (113)
[E0:1A:35:05:47:EB][LE]> 
```

Is it possible to free the device created and then use external commands during the sleep period?

### PR DESCRIPTION
**Pull request by roylee17**
Any pending reference to the socket prevents the underlying HCI device
from being released.  This most likely happens due to a blocked Read().
The Close() syscall does not unblock the Read(), and the HCI socket does
not support Shutdown() syscall, which can be used for a TCP socket.
Here we just send a HCI Reset Command, which results a HCI response that
subsequently unblocks the Read().

Attempt to fix #17 

It takes some time for kernel/driver to release the device, user might need retry & delay.
```
d, err := ble.NewDevice()
...
d.Stop()
time.Sleep(time.Millisecond * 10)
d, err = ble.NewDevice()
...
```

### CODE REVIEW

File: linux/hci/socket/socket.go [hunk 1]
```diff
@@ -131,5 +138,15 @@ func (s *Socket) Write(p []byte) (int, error) {
...
+	// Prevent subsequent Read() before we close the socket.
+	close(s.closed)
+
+	// Any pending reference to the socket prevents the underlying HCI device from being released.
+	// This most likely happens due to a blocked Read().
+	// The Close() syscall does not unblock the Read(), and the HCI socket does not support
+	// Shutdown() syscall, which can be used for a TCP socket.
+	// Here we just send a HCI Reset Command, which results a HCI response that subsequently unblocks the Read().
+	unix.Write(s.fd, []byte{0x01, 0x09, 0x10, 0x00})
+
```
Comment (moogle19) [inline linux/hci/socket/socket.go]: Could we lock the `wmu` and `rmu` here, to assure that the `Read()` and `Write()` are actually done and cannot be called during the closing of the socket?
Then `Close()` would wait until the pending `Read()` is done, then close the socket and return.

**Comment by josephroberts** [comment]
Awesome! this works nicely for me.

File: linux/hci/socket/socket.go (hunk 1 above)
Comment (roylee17) [inline linux/hci/socket/socket.go]: How about the updated one below?

`Write()` now is protected with `wmu`, and read is guarded with `rdLock`.
We're using buffer channel to emulate `TryLock()`, so `Close()` itself won't be blocked by the `rmu`.

Not sure `Close()` needs to wait for the `Read()`, but we might need the `s.closed` to notify the unblocked `Read()` that was an failed/dummy read.

Let me know anything else missing, or how we can improve, thanks.

**Comment by moogle19** [comment]
I think we should assure that `Read()` returned, before the socket is closed. This also would make the sleep between `Close()` and the creation of a new device obsolete.
If you look at the strace output, you see that the close is called before the read is resumed, which is not ideal:
```
22:41:14.753974 open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
22:41:14.754055 close(3)                = 0
22:41:14.754075 open("/usr/lib/libpthread.so.0", O_RDONLY|O_CLOEXEC) = 3
22:41:14.754092 read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\320`\0\0\0\0\0\0"..., 832) = 832
22:41:14.754175 close(3)                = 0
22:41:14.754191 open("/usr/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
22:41:14.754207 read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\260\3\2\0\0\0\0\0"..., 832) = 832
22:41:14.754275 close(3)                = 0
strace: Process 11918 attached
strace: Process 11919 attached
strace: Process 11920 attached
strace: Process 11921 attached
[pid 11917] 22:41:14.757832 ioctl(1, TCGETS, 0xc420049d84) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 11917] 22:41:14.757987 read(3, "128\n", 4096) = 4
[pid 11917] 22:41:14.758018 read(3, "", 4092) = 0
[pid 11917] 22:41:14.758036 close(3)    = 0
[pid 11917] 22:41:14.758060 socket(AF_INET, SOCK_STREAM, IPPROTO_TCP) = 3
[pid 11917] 22:41:14.758082 close(3)    = 0
[pid 11917] 22:41:14.758105 socket(AF_INET6, SOCK_STREAM, IPPROTO_TCP) = 3
[pid 11917] 22:41:14.758145 socket(AF_INET6, SOCK_STREAM, IPPROTO_TCP) = 4
[pid 11917] 22:41:14.758182 close(4)    = 0
[pid 11917] 22:41:14.758200 close(3)    = 0
[pid 11917] 22:41:14.758262 socket(AF_BLUETOOTH, SOCK_RAW, 1) = 3
[pid 11917] 22:41:14.758283 ioctl(3, HCIGETDEVLIST, 0xc420049c04) = 0
[pid 11917] 22:41:14.758305 ioctl(3, HCIDEVDOWN, 0) = 0
[pid 11917] 22:41:14.758324 ioctl(3, HCIDEVUP, 0) = 0
[pid 11917] 22:41:14.852627 ioctl(3, HCIDEVDOWN, 0) = 0
[pid 11917] 22:41:14.873732 read(3,  <unfinished ...>
[pid 11921] 22:41:14.873811 write(3, "\1\3\f\0", 4) = 4
[pid 11917] 22:41:14.877365 <... read resumed> "\4\16\4\1\3\f\0", 4096) = 7
[pid 11917] 22:41:14.877437 write(3, "\1\t\20\0", 4) = 4
[pid 11921] 22:41:14.877473 read(3, "\4\16\n\1\t\20\0\315\6\272\3011x", 4096) = 13
[pid 11921] 22:41:14.878392 read(3,  <unfinished ...>
[pid 11917] 22:41:14.878481 write(3, "\1\5\20\0", 4) = 4
[pid 11921] 22:41:14.879356 <... read resumed> "\4\16\v\1\5\20\0\375\3@\10\0\1\0", 4096) = 14
[pid 11921] 22:41:14.879403 read(3,  <unfinished ...>
[pid 11917] 22:41:14.879494 write(3, "\1\2 \0", 4) = 4
[pid 11921] 22:41:14.880348 <... read resumed> "\4\16\7\1\2 \0\33\0\17", 4096) = 10
[pid 11921] 22:41:14.880388 read(3,  <unfinished ...>
[pid 11917] 22:41:14.880473 write(3, "\1\7 \0", 4) = 4
[pid 11921] 22:41:14.881347 <... read resumed> "\4\16\5\1\7 \0\n", 4096) = 8
[pid 11921] 22:41:14.881382 read(3,  <unfinished ...>
[pid 11917] 22:41:14.881467 write(3, "\1\1 \10\37\0\0\0\0\0\0\0", 12) = 12
[pid 11921] 22:41:14.882356 <... read resumed> "\4\16\4\1\1 \0", 4096) = 7
[pid 11921] 22:41:14.882478 read(3,  <unfinished ...>
[pid 11917] 22:41:14.882528 write(3, "\1\1\f\10\377\377\373\377\7\370\277=", 12) = 12
[pid 11921] 22:41:14.883396 <... read resumed> "\4\16\4\1\1\f\0", 4096) = 7
[pid 11921] 22:41:14.883481 read(3,  <unfinished ...>
[pid 11917] 22:41:14.883595 write(3, "\1m\f\2\1\0", 6) = 6
[pid 11921] 22:41:14.884374 <... read resumed> "\4\16\4\1m\f\0", 4096) = 7
[pid 11921] 22:41:14.884447 read(3,  <unfinished ...>
[pid 11917] 22:41:14.884573 write(3, "\1\6 \17 \0 \0\0\0\0\0\0\0\0\0\0\7\0", 19) = 19
[pid 11921] 22:41:14.885350 <... read resumed> "\4\16\4\1\6 \0", 4096) = 7
[pid 11921] 22:41:14.885391 read(3,  <unfinished ...>
[pid 11917] 22:41:14.885487 write(3, "\1\v \7\1\4\0\4\0\0\0", 11) = 11
[pid 11921] 22:41:14.886351 <... read resumed> "\4\16\4\1\v \0", 4096) = 7
[pid 11921] 22:41:14.886394 read(3,  <unfinished ...>
[pid 11917] 22:41:14.886552 write(3, "\1\t\20\0", 4) = 4
[pid 11917] 22:41:14.886590 close(3)    = 0
[pid 11919] 22:41:14.886682 +++ exited with 0 +++
[pid 11920] 22:41:14.886696 +++ exited with 0 +++
[pid 11918] 22:41:14.886704 +++ exited with 0 +++
[pid 11921] 22:41:14.888480 <... read resumed> <unfinished ...>) = ?
[pid 11921] 22:41:14.889410 +++ exited with 0 +++
22:41:14.889428 +++ exited with 0 +++
```

If we add something like

```go
<-s.rdLock
defer func() { s.rdLock <- true }() // Do we need to unlock it, when the socket is closed?
```

right before the close syscall we can assure that `Read()` is done.

**Comment by roylee17** [comment]
I got your point, but it might need more complex handling I guess.

For example,  with the following you suggest (hope I didn't misunderstand it)
```
s.Write("HCI RESET")
<-s.rdLock
defer func() { s.rdLock <- true }() // Do we need to unlock it, when the socket is closed?
unix.Close()
```

But we can't guarantee that after the Write(), we'll still be the one to get the lock, right?
A reader loop() can be unblocked, and then grab the lock again.
So we'll need to write the HCI reset again (and again, until we succeed)

**Comment by moogle19** [comment]
That's true...
What if we check `closed` before the read syscall and return `io.EOF` if closed (just like the check right after the read), then we can be sure to get the lock.

**Comment by moogle19** [comment]
Just realized that my suggestion (and I think also the implementation in this PR) would block indefinitely if there is other data to read between the `s.closed = true` and the `Write()` call. The read loop would return and the write would block, wouldn't it?

**Comment by roylee17** [comment]
Ok, so I'm backing up to the earlier simple solution, which unconditionally sends a dummy HCI reset to unblock blocked Read(), if any.

This should closes #17 without side effect. We can have separate PR for improvement.

**Comment by moogle19** [comment]
Except that you have to wait x milliseconds between `Close()` and `NewDevice()`.
What about [this method](https://github.com/moogle19/ble/commit/deb6ce846401ccf0b20567ed8cdb1540d3fa000a) for closing the socket? This would prevent that another reader loop would aquire the read lock before `Close()`.

**Comment by roylee17** [comment]
@moogle19 LGTM. Open an PR with some context in the comment or log, and we merge that one instead of this. :)

**Comment by roylee17** [comment]
Superseded by #32

**Pull request closed without merging**